From a8fb49dbe0be76b0c0e7af044c4122af82d8567e Mon Sep 17 00:00:00 2001
From: Cole Robinson <crobinso@redhat.com>
Date: Tue, 20 Feb 2018 11:27:55 -0500
Subject: [PATCH 44/45] xmlapi: Fix some pylint

---
 virtinst/xmlapi.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/virtinst/xmlapi.py b/virtinst/xmlapi.py
index 270b9ebb..e4a1fe4c 100644
--- a/virtinst/xmlapi.py
+++ b/virtinst/xmlapi.py
@@ -53,7 +53,7 @@ class _XPathSegment(object):
                 self.condition_prop = cprop.strip("@")
                 self.condition_val = cval.strip("'")
             elif cond.isdigit():
-                self.condtion_num = int(cond)
+                self.condition_num = int(cond)
 
         self.is_prop = self.nodename.startswith("@")
         if self.is_prop:
-- 
2.16.2

