--- virt-manager-4.0.0/setup.py~	2022-02-12 16:13:59.000000000 +0000
+++ virt-manager-4.0.0/setup.py	2022-04-01 15:34:51.954387833 +0000
@@ -427,8 +427,8 @@
     ]
 
     def __init__(self, *args, **kwargs):
-        self.no_update_icon_cache = False
-        self.no_compile_schemas = False
+        self.no_update_icon_cache = True
+        self.no_compile_schemas = True
         setuptools.dist.Distribution.__init__(self, *args, **kwargs)
 
 
